[Cire] Step-5 구현 : 게시글 수정, 삭제 및 권한 부여 구현

cire0304 opened the pull request:
# 아쉬운 점 1

인터페이스 설계를 너무 안일하게 했다는 생각이 들었습니다.
그렇게 생각한 이유는 2가지가 있습니다.

## null -> Optional 으로 변경
당연하게 인터페이스를 변경했으니, 그에 따른 구현체들의 코드도 변경해줘야 했습니다. (너무 당연...)

## DTO를 사용하기로 결정한 이후의 코드 수정 필요
이것도 너무 당연하지만, DTO를 사용하기로 했으니 기존의 코드를 변경하거나 새로운 메서드를 만들어줘야 했습니다.

현재의 코드는 너무나도 간단하고, 양이 적어 망정이었습니다.
만약 대형 프로젝트에서 이런 안일한 인터페이스를 사용하여 구현하였다면, 대참사가 발생했을 것입니다...

앞으로는 1) 인터페이스를 생각 좀 하고 설계하거나, 2) 구현체를 먼저 만들고, 나중에 필요하면 인터페이스를 설계하는 것이 좋다는 생각을 하였습니다.

# 아쉬운 점 2

Optional을 사용함으로써, 1) Optional 안의 객체를 꺼내서 사용하거나, 2) 예외를 처리하는 부분 때문에, 코드가 너무 지저분해지는 것 같습니다.

생각해보면, 굳이 Optional에서 객체를 꺼내지 말고, 스트림을 사용하여 Optional 안의 객체를 사용하는게 더 좋은 방법인 것같습니다.
추후 리펙토링을 다시 해야할 것 같습니다

cire0304 committed:
HTML :  로그인한 사용자만 게시글 세부내용 조회 가능 하도록 코드 변경

cire0304 committed:
Config : data 및 schema 변경

cire0304 committed:
Feat : 게시글 검증 기능 업데이트

로그인된 사용자만 게시글 상세보기와 질문하기를 할 수 있음

로그인되지 않은 사용자는 로그인 하도록 했음

cire0304 committed:
Update : 게시글 작성 기능 업데이트

로그인된 사용만이 게시글 작성 가능하도록 업데이트

cire0304 committed:
Update : 게시글 클래스 관련 업데이트

TimeStamp 기능의 의존성을 낮추기 위해, 게시글 클래스의 timestamp 관련 변수를 제거하였음

cire0304 committed:
Feat : 게시글 수정 기능 추가

로그인한 사용자는 게시글을 수정할 수 있음

자신의 글만 수정할 수 있도록 추가 업데이트 예정

cire0304 committed:
Feat : 게시글 삭제 기능 추가

로그인한 사용자는 게시글을 삭제할 수 있음

자신의 글만 삭제할 수 있도록 추가 업데이트 예정

cire0304 committed:
Update : 사용자 로그인 기능 업데이트

사용자의 편의성을 위해, 로그인 후 사용자가 요청했던 URL로 리다이렉트 하도록 기능 업데이트

cire0304 committed:
Update : 사용자 수정 및 삭제 기능 업데이트

자신이 작성한 게시글만 수정하거나 삭제할 수 있도록 업데이트 하였다.

cire0304 committed:
Rename : 컨트롤러 이름 변경

이름의 명확성을 위해 이름 변경

cire0304 committed:
Remove : 사용되지 않는 라이브러리 제거

cire0304 committed:
Test : 데이터베이스 스키마 및 데이터 입력 설정 변경

cire0304 committed:
Feat : 댓글 기능 구현

댓글 기능 구현을 위해 댓글 클래스, 댓글 저장소 클래스와 컨트롤러에 댓글 관련 로직을 추가하였다.

cire0304 committed:
HTML : 댓글 관련 HTML 코드 수정

댓글을 출력 하기위한 HTML 파일을 만드는 중

cire0304 committed:
Config : 데이터베이스 스키마 및 데이터 입력 설정 변경

cire0304 committed:
Fix : 저장소 댓글 찾기 기능 버그 수정

문법 오류로 인한 버그 수정

cire0304 committed:
HTML : 댓글 관련 HTML 파일 수정

cire0304 committed:
Feat : 요청 로그 출력 인터셉터 추가

클라이언트로터 온 요청 로그를 찍기 위한 인터셉터를 추가하였다.

cire0304 committed:
Fix : 오타 수정

잘못된 문법 및 URL 을 수정

cire0304 committed:
Feat : 댓글 추가 및 삭제 기능 추가

로그인한 사용자만이 댓글을 추가할 수 있고, 자신이 작성한 댓글만 삭제할 수 있다.

honux77 commented on src/main/java/kr/codesqaud/cafe/filter/LoginCheckInterceptor.java:
```diff
@@ -0,0 +1,21 @@
+package kr.codesqaud.cafe.filter;
```
패키지 이름이 필터인데 필터와 인터셉터는 꽤 다른 건데요. 정확한 차이점을 공부해 주셔도 좋을 것 같아요.

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -20,14 +20,20 @@ public class ArticleController {
+    }
 
+    @GetMapping("/createForm")
+    public String create(@SessionAttribute(name = ConstConfig.SESSION_ID) User user, Model model) {
```
```suggestion
    public String create(@SessionAttribute(name = ConstantConfig.SESSION_ID) User user, Model model) {
```

가능하면 줄임말보다는 원래의 단어가 좋다고 생각합니다.

honux77 commented on src/main/java/kr/codesqaud/cafe/repository/h2Repository/H2ArticleRepository.java:
```diff
@@ -25,19 +25,26 @@ public H2ArticleRepository(DataSource dataSource) {
     }
 
-    public Article findByIndex(int index) {
+    public Article findByArticleId(int index) {
```
```suggestion
    public Optional<Article> findByArticleId(int index) {
```
결과가 없을때 null보다는 Optional이겠죠?

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -39,13 +45,46 @@ public String list(Model model) {
     }
 
-    @GetMapping("show/{articleId}")
+    @GetMapping("/show/{articleId}")
```
게시글이 없을 때 예외처리를 하면 좋을 것 같아요.

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -39,13 +45,46 @@ public String list(Model model) {
+                         Model model) {
+        // TODO : 검증 공부 후 다시 리펙토링 할 예정
+        Article article = articleRepository.findByArticleId(articleId);
+        if (!article.getWriter().equals(user.getUserId())) {
```
현재 업데이트 권한이 없을 때 그냥 리스트로 돌아가는데 그것보다 명시적인 에러 처리를 해 주는 게 좋을 것 같아요.

honux77 reviewed (changes requested):
수고하셨습니다.
TODO 부분과 코멘트 부분 참고해서 개선해 주시면 좋겠습니다.

cire0304 committed:
Rename : 코드의 이름 변경

이름의 형식이 맞지않거나, url의 형식을 맞춰주기 위해 이름을 변경

cire0304 committed:
Refacor : 사용자 로그인 및 업데이트 기능 리펙토링

사용자와 관련된 메소드의 입력 형식을 규격화 시키기위해, DTO를 이용하기위한 리팩토링을 진행하였다.

cire0304 committed:
Config : 데이터베이스 변경

공부를 위해 H2 데이터베이스에서 MySQL로 데이터베이스를 변경하였다.

cire0304 committed:
Rename : HTML 에서 URL을 변경

변경하지 않는 url 변경

cire0304 committed:
Rename : 데이터베이스 관련 클래스 이름 변경

특정 데이터베이스 이름이 아닌, 추상화된 데이터베이스 이름을 사용하였다.
왜냐하면, DI를 통해, 또는 데이터 베이스 Driver의 변경으로 데이터베이스를 쉽게 바꿀 수 있기 때문이다.

cire0304 committed:
Refactor : 데이터 검색 로직 코드 리펙토링

데이터베이스에서 데이터를 찾았을 때, 찾지 못할 경우를 커버하기 위해, Optional을 사용하였다.

cire0304 committed:
Rename : Const 클래스 이름 변경

줄임말을 쓰지 않도록 이름을 변경하였다.

honux77 commented:
@dltpwns0 이거 머지해도 되나요??
